src/query_router.rs

```diff
@@ -256,23 +261,26 @@ impl QueryRouter {
     }
 
     /// Try to infer which server to connect to based on the contents of the query.
-    pub fn infer(&mut self, mut buf: BytesMut) -> bool {
+    pub fn infer(&mut self, buf: &BytesMut) -> bool {
         debug!("Inferring role");
 
-        let code = buf.get_u8() as char;
-        let len = buf.get_i32() as usize;
+        let mut message_cursor = Cursor::new(buf);
+
+        let code = message_cursor.get_u8() as char;
+        let _len = message_cursor.get_i32() as usize;
 
         let query = match code {
             // Query
             'Q' => {
-                let query = String::from_utf8_lossy(&buf[..len - 5]).to_string();
+                let query = message_cursor.read_string().unwrap();
```

Comment:
Here we should really read the exact amount we know is in the message instead of reading until a NULL byte.